### PR TITLE
Add Block Strings Support

### PR DESCRIPTION
**Pull request by DimuthuMadushan**
## Purpose

Fixes: [#1704](https://github.com/ballerina-platform/ballerina-standard-library/issues/1704)

## Examples
In GraphQL documents, when we need multi-line strings, we can use block strings as follow:
```graphql
query sendEmail() {
    sendEmail(message: """
        Hello,
          World!,

        Yours,
            GraphQL
    """)
}
```
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina/variable_validator.bal [hunk 1]
```diff
@@ -128,13 +128,29 @@ class VariableValidator {
         if argument.getKind() == parser:T_IDENTIFIER {
             argument.setValue(value);
         } else if getTypeNameFromValue(value) == getTypeName(argument) {
-            argument.setValue(value);
+            if getTypeNameFromValue(value) == STRING {
+                if self.hasInvalidCharaters(<string>value) {
```
Comment (ayeshLK) [inline ballerina/variable_validator.bal:132]: Shall we use `value:ensureType(v, string)` to safely cast variables ?

File: ballerina/variable_validator.bal [hunk 2]
```diff
@@ -128,13 +128,29 @@ class VariableValidator {
...
-            self.errors.push(getErrorDetailRecord(message, location)); 
+            self.errors.push(getErrorDetailRecord(message, location));
+        }
+    }
+
+    public isolated function hasInvalidCharaters(string value) returns boolean {
+        if value.includes(string`"""`) || value.includes(string`\"""`)  {
+            return true;
         }
+        return false;
```
Comment (ayeshLK) [inline ballerina/variable_validator.bal:154]:
```suggestion
    public isolated function hasInvalidCharaters(string value) returns boolean {
        return value.includes(string`"""`) || value.includes(string`\"""`);
```

File: ballerina/variable_validator.bal (hunk 1 above)
Comment (DimuthuMadushan) [reply ballerina/variable_validator.bal:132]: Since we have checked the value at line 131, Could it be a problem use direct casting here?

File: ballerina/modules/parser/lexer.bal [hunk 3]
```diff
@@ -103,22 +103,62 @@ public class Lexer {
...
-                break;
+            if value == DOUBLE_QUOTE {
+                previousChar = self.readNextChar();
+                value = self.charReader.peek();
+                if value == DOUBLE_QUOTE {
+                    previousChar = self.readNextChar();
+                    value = self.charReader.peek();
+                    if value == DOUBLE_QUOTE {
+                        value = self.readNextChar();
+                        break;
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:150]: There are two places where we have to check for the triple-double quotes. Isn't it be easy if we use a separate function for this?

**Comment by ThisaruGuruge** [comment]
There's a test failure in windows build too.

File: ballerina/modules/parser/lexer.bal [hunk 4]
```diff
@@ -103,22 +103,62 @@ public class Lexer {
...
         _ = self.readNextChar(); // Ignore first double quote character
-        while (true) {
+        string value = self.charReader.peek();
+        if value == DOUBLE_QUOTE {
+            value = self.readNextChar();
+            value = self.charReader.peek();
+            if value == DOUBLE_QUOTE {
+                return self.readBlockStringLiteral();
+            } else {
+                return  getToken(word, T_STRING, location);
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply ballerina/modules/parser/lexer.bal:113]: Shall we add a test containing an empty string (`""`) for this too?

File: ballerina/modules/parser/lexer.bal [hunk 5]
```diff
@@ -103,22 +103,112 @@ public class Lexer {
...
+                lines.push(word);
+                word = "";
                 value = self.readNextChar();
-                break;
+            } else if value == DOUBLE_QUOTE {
+                previousChar = self.readNextChar();
+                value = self.charReader.peek();
+                if value == DOUBLE_QUOTE {
+                    previousChar = self.readNextChar();
+                    value = self.charReader.peek();
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:153]: Do we have a test for this?
i.e. a case where `""` occurs in a block string? If not shall we add one?

File: ballerina/modules/parser/lexer.bal [hunk 6]
```diff
@@ -103,22 +103,112 @@ public class Lexer {
...
+    isolated function getBlockStringValue(string[] lines) returns string {
+        int? commonIndent = ();
+        string formatted = "";
+        foreach string line in lines {
+            int indent  = self.getLeadingWhiteSpaceCount(line);
+            if indent < line.length() && (commonIndent is () || indent < commonIndent) {
+                commonIndent = indent;
+                if commonIndent == 0 {
+                    break;
+                }
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:181]: Is this necessary? This is a special case right? Do we need to consider the special case here, since this will be handled anyways?

File: ballerina/tests/24_block_strings.bal [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+@test:Config {
+    groups: ["block_strings"]
+}
+isolated function testBlockStringsWithVariableDefaultValue() returns error? {
+    http:Request request = new;
+    string document = check getGraphQLDocumentFromFile("block_string_with_variable_default_value.txt");
+    string url = "http://localhost:9091/inputs";
+    json actualPayload = check getJsonPayloadFromService(url, document);
+    json expectedPayload = check getJsonContentFromFile("block_strings_with_variables_default_value.json");
```
Comment (ThisaruGuruge) [inline ballerina/tests/24_block_strings.bal:40]: Name mismatch?
```suggestion
    json expectedPayload = check getJsonContentFromFile("block_string_with_variables_default_value.json");
```

File: ballerina/tests/resources/documents/block_strings_with_invalid_characters.txt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+    sendEmail(message: """Hello,
+            World!,
+\"""
+        Yours,
+            GraphQL""") 
+}
```
Comment (ThisaruGuruge) [inline ballerina/tests/resources/documents/block_strings_with_invalid_characters.txt:7]: What is the invalid character in this?

File: ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
+    "errors": [
+        {
+            "message": "Syntax Error: Cannot parse the unexpected character \"\"\".",
```
Comment (ThisaruGuruge) [inline ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json:4]: Why does this consider an invalid character?

**Comment by ThisaruGuruge** [comment]
There are still test failures in Windows.

File: ballerina/modules/parser/lexer.bal (hunk 5 above)
Comment (DimuthuMadushan) [reply ballerina/modules/parser/lexer.bal:153]: We already have a test named ```testBlockStringsWithDoubleQuotes```.

File: ballerina/modules/parser/lexer.bal (hunk 6 above)
Comment (DimuthuMadushan) [reply ballerina/modules/parser/lexer.bal:181]: will remove it

File: ballerina/tests/resources/documents/block_strings_with_invalid_characters.txt (hunk 8 above)
Comment (DimuthuMadushan) [reply ballerina/tests/resources/documents/block_strings_with_invalid_characters.txt:7]: This test name should be ```testInvalidBlockStrings```. I will change it.

File: ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json (hunk 9 above)
Comment (DimuthuMadushan) [reply ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json:4]: This error is not coming from the block string reading. Since this input block string includes a ```""""``` in the middle, rest of the string consider as an identifier. Then this first ```"``` at the end of the the string consider as an invalid character since identifier can't have ```"``` as a character.

Comment (DimuthuMadushan) [reply ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json:4]: @ThisaruGuruge is this the correct scenario?

Comment (ThisaruGuruge) [reply ballerina/tests/resources/expected_results/block_strings_with_invalid_characters.json:4]: I still don't understand why is this considered an invalid character (or character sequence).

IMHO, any character sequence is accepted within the block string, except the `"""`, which terminates the block.

Even if you try the same block string in the Appollo server, it does not recognize this as an invalid character. What's the basis of considering this an invalid character?

File: ballerina/modules/parser/lexer.bal [hunk 10]
```diff
@@ -103,22 +103,124 @@ public class Lexer {
...
                 value = self.readNextChar();
-                break;
+            } else if value == DOUBLE_QUOTE {
+                string wordWithQuotes = self.readQuotesInBlockStrings(word);
+                if wordWithQuotes == word {
+                    lines.push(word);
+                    break;
+                } else {
+                    word = wordWithQuotes;
+                }
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:155]:
```suggestion
                word = wordWithQuotes;
```

We are pushing the line anyways if we find a new line character, right?

File: ballerina/modules/parser/lexer.bal [hunk 11]
```diff
@@ -103,22 +103,124 @@ public class Lexer {
...
                 word += value;
             }
             previousChar = value;
         }
-        return getToken(word, T_STRING, location);
+        return self.getBlockStringValue(lines);
+    }
+
+    isolated function readQuotesInBlockStrings(string word) returns string {
+        string previousChar = self.readNextChar();
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:166]: This variable is redundant. It has only been used to concatenate the character inside the loop. The character is not carried forward for the next iteration. We can use a local variable for that.

File: ballerina/modules/parser/lexer.bal [hunk 12]
```diff
@@ -103,22 +103,124 @@ public class Lexer {
...
+        int? commonIndent = ();
+        string formatted = "";
+        foreach string line in lines {
+            int indent  = self.getLeadingWhiteSpaceCount(line);
+            if indent < line.length() && (commonIndent is () || indent < commonIndent) {
+                commonIndent = indent;
+            }
+        }
+        if commonIndent is int {
+            foreach string line in lines {
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:197]: Since we need the index of the line inside the loop, better to use a for loop with a number range instead of a foreach loop. 
[Check this BBE](https://ballerina.io/learn/by-example/foreach-statement.html?is_ref_by_example=true)

File: ballerina/modules/parser/lexer.bal [hunk 13]
```diff
@@ -103,22 +103,124 @@ public class Lexer {
...
+                if commonIndent > line.length() {
+                    continue;
+                } else {
+                    lines[<int>lines.indexOf(line)] = line.substring(commonIndent, line.length());
+                }
+            }
+        }
+        foreach string line in lines {
+            formatted += line + "\n";
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/lexer.bal:206]: Use [join](https://lib.ballerina.io/ballerina/lang.string/1.1.0/functions#'join) instead. 

```suggestion
            formatted += string:'join("\n", lines);
```

File: ballerina/tests/24_block_strings.bal (hunk 7 above)
Comment (ThisaruGuruge) [reply ballerina/tests/24_block_strings.bal:37]: Still, the names does not match :) 

```
block_strings_with_variable_default_value.txt
block_strings_with_variables_default_value.json
```

File: ballerina/tests/resources/expected_results/invalid_block_strings.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
+{
+    "errors": [
+        {
+            "message": "Syntax Error: Cannot parse the unexpected character \"\"\".",
```
Comment (ThisaruGuruge) [inline ballerina/tests/resources/expected_results/invalid_block_strings.json:4]: This error seems to be not right.  In the given document, the block string ends at `{ line: 4, column: 1 }`.
Then there's another identifier, which means we have another argument. Then there's another identifier without a colon character, which is incorrect.

So the error should be 
```
{
    message: "Syntax Error: Expected \":\", found Name.",
    locations: [
        {
            line: 6,
            column: 14
        }
    ]
}
```

File: ballerina/modules/parser/lexer.bal [hunk 15]
```diff
@@ -103,22 +103,124 @@ public class Lexer {
...
-                string message = "Syntax Error: Unterminated string.";
-                return error UnterminatedStringError(message, line = self.currentLocation.line,
-                                                     column = self.currentLocation.column);
-            } else if (value == DOUBLE_QUOTE && previousChar != BACK_SLASH) {
+            if value == NEW_LINE {
+                lines.push(word);
+                word = "";
                 value = self.readNextChar();
-                break;
+            } else if value == DOUBLE_QUOTE {
```
[the remark this answers is not in the document]
Comment (DimuthuMadushan) [reply ballerina/modules/parser/lexer.bal:148]: if we use `else if value == DOUBLE_QUOTE && previousChar != BACK_SLASH {
`, we should use `else if value == DOUBLE_QUOTE && previousChar == BACK_SLASH {
` to remove the `\` in this special case `\" " "`. This may repeat the same thing.

[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply ballerina/modules/parser/lexer.bal:148]: No, the point is to identify the end of the block string, which is `value == DOUBLE_QUOTE && previousChar != BACK_SLASH`, and then two consecutive `"` characters. Otherwise, it is just another character.
